safetensors/src/slice.rs

```diff
@@ -0,0 +1,416 @@
+        let mut indices = vec![];
+        // Everything is row major.
+        for (i, &shape) in view.shape.iter().enumerate().rev() {
+            if i >= slices.len() {
+                // We are  not slicing yet, just increase the local span
+                span *= shape;
+            } else {
+                let slice = &slices[i];
+                let (start, stop) = match slice {
+                    TensorIndexer::Narrow(Bound::Unbounded, Bound::Unbounded) => (0, shape),
+                    TensorIndexer::Narrow(Bound::Unbounded, Bound::Excluded(stop)) => (0, *stop),
+                    TensorIndexer::Narrow(Bound::Unbounded, Bound::Included(stop)) => {
+                        (0, *stop + 1)
+                    }
+                    TensorIndexer::Narrow(Bound::Included(s), Bound::Unbounded) => (*s, shape),
+                    TensorIndexer::Narrow(Bound::Included(s), Bound::Excluded(stop)) => (*s, *stop),
+                    TensorIndexer::Narrow(Bound::Included(s), Bound::Included(stop)) => {
+                        (*s, *stop + 1)
+                    }
+                    TensorIndexer::Narrow(Bound::Excluded(s), Bound::Unbounded) => (*s + 1, shape),
+                    TensorIndexer::Narrow(Bound::Excluded(s), Bound::Excluded(stop)) => {
+                        (*s + 1, *stop)
+                    }
+                    TensorIndexer::Narrow(Bound::Excluded(s), Bound::Included(stop)) => {
+                        (*s + 1, *stop + 1)
+                    }
+                };
+                if indices.is_empty() {
+                    if start == 0 && stop == shape {
+                        // We haven't started to slice yet, just increase the span
+                        span *= shape;
+                    } else {
+                        let offset = start * span;
+                        span = stop * span - offset;
+                        indices.push((offset, offset + span));
+                        span *= shape;
+                    }
+                } else {
+                    let mut newindices = vec![];
+                    for n in start..stop {
+                        let offset = n * span;
+                        for (old_start, old_stop) in &indices {
+                            newindices.push((old_start + offset, old_stop + offset));
+                        }
+                    }
+                    indices = newindices;
+                }
+            }
+        }
+        if indices.is_empty() {
+            indices.push((0, view.data.len()));
+        }
+        // Reversing so we can pop faster while iterating on the slice
+        let indices = indices.into_iter().rev().collect();
+        Ok(Self { view, indices })
+    }
+}
+
+impl<'data> Iterator for SliceIterator<'data> {
+    type Item = &'data [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // TODO
+        let (start, stop) = self.indices.pop()?;
+        Some(&self.view.data[start..stop])
```

Comment:
Because we calculate all indices in advance this becomes extremely simple.

I think this is OK in first approximation, but for iterators it's probably nicer if we can avoid the full calculation and storage upfront and implement only the next logic.

This seems simpler for now. We will fix if there is actually a performance issue I think